### ISSUE TITLE
ncurses/ncurses.h not found in copiling/linking

### ISSUE THREAD

**Issue by hygitter** [nezticle/RaspberryPi-BuildRoot#35]
Hi! I was trying to compile a bsqask image for raspberry-pi with the step by step guide (that is running the toolchain?). After about 1,5 hours i got the error

```
/opt/bsquask/build/host-cmake-2.8.10.2/Source/CursesDialog/cmCursesStandardIncludes.h:32:31: Schwerwiegender Fehler: ncurses/ncurses.h: Datei oder Verzeichnis nicht gefunden
Kompilierung beendet.
```

(fatal error, ncurses/ncurses.h: file not found) - More: http://pastebin.com/yXViQhzR

I was using ubuntu 12.10, and used the following commands, also to install missing software which i was told to install during the process. see: http://pastebin.com/jXJSEbNs

I've no clue what this might mean, and would be delighted by every hint. This is the bsquask branch, here, I also (first) tried the master (normal) branch exactly like in the readme, with the same result.

Cheers, hygitter

Edit: I also got a Application-has-an-error window from ubuntu, saying send-an-error report? Clicking yes says the program ... causing the error is no linger installed. ( I forgot the program name but I can find out if I run it once again)
Edit2: It is  /opt/bsquask/build/host-m4-1.4.16/conftest

**Comment by nezticle** [comment]
that is an error I've not encountered before.  The problem is that when you are building the host cmake (the one that is run locally to build cmake based projects), it can't find the ncurses development files. I would think that this would be installed for you, but it doesn't seem that is the case (and would be a problem in upstream BuildRoot as well)

So as a solution I see two options:

1) try to manually install the "host-ncurses" package in the bsquask repo.  So instead of just calling make in your bsquask build folder call "make host-ncurses" instead, then when that finishes, run "make" again.

2) If that doesn't work then install the ncurses development files on your local machaine.  Technically you are building cmake to be run locally, so maybe this assumes that you have the ncurses development files installed.  To install this on ubuntu, just run "sudo apt-get install ncurses-dev"
Then just run "make" again to continue

Also, let me know if that works, because I can fix the build script for other in the future.

**Comment by hygitter** [comment]
Super Thanks! I'll try this, number 1 first. Starting tomorrow or so.
Bye!

**Comment by pepeki** [comment]
I had the same problems while building Bsquask on a clean Ubuntu 12.10 (with xwindows, kde, ... removed) recently...
I tried installing ncurses - which had no effect...
I ended up manually modifying includes of the files in trouble (3-4 files) - excluding the relative path...

Peter

**Comment by hygitter** [comment]
Ok number 1) ends up like this: http://pastebin.com/zY9XGi9z

```
Invalid configuration `i686-pc-linux-': machine `i686-pc-linux' not recognized
configure: error: /bin/bash ./config.sub i686-pc-linux- failed.
```

now trying number 2)
Cheers, hygitter

**Comment by ghost** [comment]
why are you using 'sudo' ?
The complete buildroot can (should?) be used as a normal user. I don't know if this has anything to do with your build error, but everything can be done as a normal user, no root privilege necessary.

**Comment by hygitter** [comment]
Number 2): installed  `ncurses-dev`, deleting /opt/bsquask, did make raspberrypi_defconfig like in the readme file ... and did `sudo make host-ncurses`, just a guess to try for now. with the same result like above, `Invalid configuration i686-pc-linux-: machine i686-pc-linux not recognized`. I'll try `make` later on, and I also like pepekis way to look in the files, thank you!

Cheers hybroke

@limpens I can't write /opt in normal mode: trying 

```
mkdir /opt/rr
mkdir: kann Verzeichnis »/opt/rr“ nicht anlegen: Keine Berechtigung
```

gives me not allowed. Should I build somewhere else? in ~?  
bye

**Comment by ghost** [comment]
I'ld suggest you build somewhere else, to avoid breaking privileges on other directories.
in my case, I've cloned the repo in ~/dev/rpi/RaspberryPi-Buildroot/ and don't need escalated privileges. but any directory in your home directory would do.
If there is nothing else using to /opt directory, you could ofcourse change the owner/group of the directory, but this might break things (now or in the future).

**Comment by qdotme** [comment]
Actually, it seems that it needs host's (`apt-get install ncurses-dev`) ncurses, but with a twist:

```
ln -s /usr/include /usr/include/ncurses 
```

Since the cmake expects `ncurses.h` in `/usr/include/ncurses/ncurses.h`, while `ncurses-dev` installs in `/usr/include/ncurses.h`

Tested on Ubuntu 12.04.2 Server

**Comment by hygitter** [comment]
~.~.~Yess! I tried nezticle's way 2) and also qdotdot's symlink and also installed something else which I found somewhere and it worked. Unfortunately now I don't know if it would work without symlink, probably it would work without the other installs I found. Here are my additional commands:

```
sudo apt-get install ncurses-dev
ln -s /usr/include /usr/include/ncurses 
sudo apt-get install git-core subversion
```

I kept building in /opt/bsquask for now. I didn't get an Application-has-an-error window from ubuntu, something about conftest, like before.

Thank you all! #gitter

**Comment by nezticle** [comment]
This is a legitimate error that I can reproduce with a fresh install of Ubuntu 13.04.  It seems that the cmake build gets confused and adds $BSQUASK_HOST_DIR/include/ncurses/ to its INCLUDE path, but then still includes the headers like "#include "ncurses/ncurses.h" which won't work.  Ideally it would add $BSQUASK_HOST_DIR/include/ instead to its INCLUDE path.  I'm investigating solutions now (as this works fine on my 12.04 installation without using the local system headers)

**Comment by nezticle** [comment]
``` c
/* figure out which curses.h to include */
# if defined(CURSES_HAVE_NCURSES_H)
#  include <ncurses.h>
# elif defined(CURSES_HAVE_NCURSES_NCURSES_H)
#  include <ncurses/ncurses.h>
# elif defined(CURSES_HAVE_NCURSES_CURSES_H)
#  include <ncurses/curses.h>
# else
#  include <curses.h>
# endif
```

When:

```
CURSES_HAVE_NCURSES_CURSES_H:FILEPATH=/home/nezticle/Bsquask/host/usr/include/ncurses/curses.h
CURSES_HAVE_NCURSES_NCURSES_H:FILEPATH=/home/nezticle/Bsquask/host/usr/include/ncurses/ncurses.h
CURSES_INCLUDE_PATH:FILEPATH=/home/nezticle/Bsquask/host/usr/include/ncurses
CURSES_LIBRARY:FILEPATH=/home/nezticle/Bsquask/host/usr/lib/libncurses.a
CURSES_NCURSES_LIBRARY:FILEPATH=/home/nezticle/Bsquask/host/usr/lib/libncurses.a
cmForm_LIB_DEPENDS:STATIC=general;/home/nezticle/Bsquask/host/usr/lib/libncurses.a;
```

**Comment by nezticle** [comment]
Changing the variable:

```
//The curses include path
CURSES_INCLUDE_PATH:FILEPATH=/home/nezticle/Bsquask/host/usr/include
```

In CMakeCache.txt after configure works around the problem, so this is definately the case, but that means that the logic in Modules/FindCurses.cmake is broken for the BuildRoot environment, as it sets the CURSES_INCLUDE_PATH variable incorrectly.

**Comment by nezticle** [comment]
Fixed: f2c7f0b4bf9e298f95fa114b77ede4bc9b2712b0
